jzaia_files/main.cpp

```diff
@@ -0,0 +1,291 @@
+        GateImplementationsLM::applyHadamard(sv.getData(),
+                                             sv.getNumQubits(),
+                                             {i},
+                                             false);
+    }
+}
+
+/**
+ * @brief Apply the "Grover's Mirror" reflection to the given statevector
+ *
+ * Performs a reflection across the vector which represents the uniform
+ * superposition. This is used for amplitude-amplification for Grover's
+ * algorithm.
+ *
+ * @param sv The statevector to reflect
+ */
+void groversMirror(StateVectorLQubitManaged<double> &sv) {
+    const size_t num_qubits = sv.getNumQubits();
+
+    // Apply H to all non-ancilla qubits
+    for (size_t i = 0; i < num_qubits-1; ++i) {
+        GateImplementationsLM::applyHadamard(sv.getData(),
+                                             sv.getNumQubits(),
+                                             {i},
+                                             false);
+    }
+
+    std::vector<size_t> controls(num_qubits-1);
+    std::vector<bool> control_values(num_qubits-1, false);
+    std::iota(controls.begin(), controls.end(), 0);
+
+    // Apply an anti-controlled on all search qubits X-gate onto the ancilla qubit
+    GateImplementationsLM::applyNCPauliX(sv.getData(),
+                                         sv.getNumQubits(),
+                                         controls,
+                                         control_values,
+                                         {num_qubits-1},
+                                         false);
+
+    // Apply H to all non-ancilla qubits
+    for (size_t i = 0; i < num_qubits-1; ++i) {
+        GateImplementationsLM::applyHadamard(sv.getData(),
+                                             sv.getNumQubits(),
+                                             {i},
+                                             false);
+    }
+}
+
+/**
+ * @brief The first testing oracle for Grover's
+ *
+ * A 6-qubit test oracle for Grover's algorithm. Applies a pauli-X to the
+ * rightmost qubit if the leftmost 5 qubits are in the state |11010>
+ *
+ * @param sv The statevector to apply the oracle to. Must be 6 qubits
+ */
+void oracle1(StateVectorLQubitManaged<double> &sv) {
+    // Sanity check statevector
+    assert(sv.getNumQubits() == ORACLE1_QUBITS);
+
+    // Define controls to be used for applying the X gate
+    std::vector<size_t> controls(ORACLE1_QUBITS-1);
+    std::iota(controls.begin(), controls.end(), 0);
+    std::vector<bool> control_vals = ORACLE1_EXPECTED;
+
+    // Apply the X gate to the ancilla, controlled on the chosen bitstring
+    GateImplementationsLM::applyNCPauliX(sv.getData(),
+                                         sv.getNumQubits(),
+                                         controls,
+                                         control_vals,
+                                         {ORACLE1_QUBITS-1},
+                                         false);
+}
```

Comment:
Can you think of a way to have a function Oracle defined a single time and that we could re-use for the three test cases?

Reply:
Yes, I originally didn't do this so I could support more complex oracles that might not follow the same pattern (for example, one which selects multiple states by using multiple smaller controlled nots rather than just doing a single all-qubit controlled not). After some discussion in the github issue thread, I decided not to pursue other oracles, but I left this pattern as such. Since I made the python version later on, I used a function which generates other functions there to condense everything. Something similar could be done in the C++ version using preprocessor commands, or even more simply by just having a single oracle function which takes the desired control qubits as a parameter.